### PR TITLE
Fix the maximization of sub windows

### PR DESCRIPTION
**Pull request by michaelgregorius**
Pull request https://github.com/LMMS/lmms/pull/7514 more or less consists of two fixes. The first one fixes the rendering of maximized `SubWindow` instances and the second one enables instrument windows to be maximizable. This pull request separates out the first fix into its own pull request.

Technically its the same pull request as https://github.com/LMMS/lmms/pull/7514 but commit 9a1eaac8ec1 has been reverted in this branch.

### CODE REVIEW

**Comment by JohannesLorenz** [comment]
During testing I noted (also with my own resizable PR) this issue:

1. Start LMMS with the default project where you have TripleOsc
2. Drag SlicerT onto the TripleOsc TrackLabelButton
3. Click the TrackLabelButton to open SlicerT

When you do this: SlicerT seems to be not resizable.

File: src/gui/SubWindow.cpp [hunk 1]
```diff
@@ -111,6 +111,13 @@ SubWindow::SubWindow(QWidget *parent, Qt::WindowFlags windowFlags) :
  */
 void SubWindow::paintEvent( QPaintEvent * )
 {
+	// Don't paint any of the other stuff if the sub window is maximized
```
Comment (JohannesLorenz) [inline src/gui/SubWindow.cpp:98]: The `p.fillRect( rect, isActive ? activeColor() : p.pen().brush() );` does not look like "other stuff", but rather like background which should also be required on maximized windows? Can you please elaborate why this is not needed for maximized windows?

**Comment by JohannesLorenz** [comment]
The following issue was already in the SlicerT update, but could be fixed here:

1. Open LMMS
2. Create a Zyn track
3. Open Zyn by pressing the TrackLabelButton

The Zyn window will be too big. The following fixes it:

```patch
diff --git a/src/gui/instrument/InstrumentTuningView.cpp b/src/gui/instrument/InstrumentTuningView.cpp
index daa361aad..d495eb98a 100644
--- a/src/gui/instrument/InstrumentTuningView.cpp
+++ b/src/gui/instrument/InstrumentTuningView.cpp
@@ -64,7 +64,7 @@ InstrumentTuningView::InstrumentTuningView(InstrumentTrack *it, QWidget *parent)
        masterPitchLayout->addWidget(tlabel);
 
        // Microtuner settings
-       m_microtunerNotSupportedLabel = new QLabel(tr("Microtuner is not available for MIDI-based instruments."));
+       m_microtunerNotSupportedLabel = new QLabel(tr("Microtuner is not available\nfor MIDI-based instruments."));
        m_microtunerNotSupportedLabel->setWordWrap(true);
        m_microtunerNotSupportedLabel->hide();
        layout->addWidget(m_microtunerNotSupportedLabel);
```

**Comment by JohannesLorenz** [comment]
Lv2 instruments are resizable to 0 size if I rebase lv2-ui to your branch (I did this on my fork, branch `lv2-ui-without-resize`). However, resizing the Instrument window shall not allow to make it that small that the buttons in the tab widget are pushed together, or that any of the tab widgets children gets too small for its contents. See especially the TabWidget fixes in 246971a5d0d4c9ce31c9b5d556726e60c1d4787a (branch lv2-ui).

**Comment by JohannesLorenz** [comment]
Drag the preset "Kicker/SnareMarch.xpf" into the Song Editor, then open the Instrument (#5851). Observartion: #5851 is not observed anymore, but the effects in the InstrumentWindow are "too short". I must still check whether and if yes, how I fixed this. Likely introduced by the SlicerT update.

**Comment by JohannesLorenz** [comment]
> When you do this: SlicerT seems to be not resizable.

The opposite way around also does not work: TripleOsc is resizable then (it can get too small and too large).

**Comment by JohannesLorenz** [comment]
I think that's it for now! If you can fix all this, I am happy :laughing: Anyways, so far, good work, I could not find too many issues.

File: src/gui/SubWindow.cpp (hunk 1 above)
Comment (michaelgregorius) [reply src/gui/SubWindow.cpp:98]: The `rect` is the rectangle of the title bar which should not be rendered when the window is maximized. It is defined a few lines before the `fillRect` call:
```
QRect rect( 0, 0, width(), m_titleBarHeight );
```

Painting it when the window is maximized led to the following buggy rendering where the title bar is rendered behind the main content of the sub window:

7512-ArtifactsForMaximizedInstrumentWindow

When the window is maximized we only want to render the child content and therefore we exit early if the window is maximized.

**Comment by michaelgregorius** [comment]
@JohannesLorenz, it seems to me that most of the problems that you have listed should not be caused by the changes in this PR as it is intended to only fix the rendering of maximized sub windows and does not introduce any capabilities for windows to maximize. Enabling to maximize instrument windows is done by the following pull request which needs to be merged with master once this PR is merged: https://github.com/LMMS/lmms/pull/7514.

Do all the problems that you have listed disappear if you rebase your branch on the current master? If that's not the case then let's please not mix up pull requests, what they do and what should be fixed before they are merged. It seems that most of your remarks should be discussed in https://github.com/LMMS/lmms/pull/7514 or similar PRs.

I have added commit 6b953fc0d58 which makes sure that all flags that have previously been set are also set in the new code, i.e. `Qt::SubWindow`, `Qt::WindowMaximizeButtonHint`, `Qt::WindowSystemMenuHint`, `Qt::WindowCloseButtonHint`, `Qt::CustomizeWindowHint`. The first four flags are already set as window flags when the code reaches the point to configure the flags, so only `Qt::CustomizeWindowHint` had to be added explicitly.

File: src/gui/SubWindow.cpp (hunk 1 above)
Comment (JohannesLorenz) [reply src/gui/SubWindow.cpp:98]: Oh, of course, that rectangle. Thanks, resolving.

**Comment by JohannesLorenz** [comment]
> Do all the problems that you have listed disappear if you rebase your branch on the current master? If that's not the case then let's please not mix up pull requests

That's totally correct. Thanks for the hint! I will remove my comments soon and will probably just append them to issue #7510 . Will continue the review and testing.

**Review by JohannesLorenz — COMMENTED** [review]
What I did:
* First functional review
* Testing review (No issues found, but note there is one issue in the related PR - not 100% sure which PR it belongs to)

What needs to be done:
* Style review
* Second functional review (IMO optional)

File: src/gui/SubWindow.cpp [hunk 2]
```diff
@@ -295,9 +284,25 @@ void SubWindow::moveEvent( QMoveEvent * event )
...
+	// Don't show the title or any button if the sub window is maximized. Otherwise they
+	// might show up behind the actual maximized content of the child widget.
+	if (isMaximized())
+	{
+		m_closeBtn->hide();
+		m_maximizeBtn->hide();
+		m_restoreBtn->hide();
+		m_windowTitle->hide();
+
+		return;
```
[the remark this answers is not in the document]
Comment (michaelgregorius) [reply src/gui/SubWindow.cpp:296]: Good catch! Fixed with commit 40a90ad7a2d.

File: src/gui/SubWindow.cpp [hunk 3]
```diff
@@ -295,9 +284,25 @@ void SubWindow::moveEvent( QMoveEvent * event )
...
+	}
+
+	// Title adjustments
+	m_windowTitle->show();
+
 	// button adjustments
 	m_maximizeBtn->hide();
```
[the remark this answers is not in the document]
Comment (michaelgregorius) [reply src/gui/SubWindow.cpp:306]: Yes, this is intentional. I wanted to make clear that the maximized state and the non-maximized state are two distinct things and did not want to mix up their code.

**Comment by michaelgregorius** [comment]
@Rossmaxx, can you do a style review as proposed by @JohannesLorenz [here](https://github.com/LMMS/lmms/pull/7530#pullrequestreview-2355543462)?

File: include/SubWindow.h [hunk 4]
```diff
@@ -78,6 +78,8 @@ class LMMS_EXPORT SubWindow : public QMdiSubWindow
 	void paintEvent( QPaintEvent * pe ) override;
 	void changeEvent( QEvent * event ) override;
 
+	QPushButton* addTitleButton(const std::string & iconName, const QString & toolTip);
```
Comment (Rossmaxx) [inline include/SubWindow.h:81]:
```suggestion
	QPushButton* addTitleButton(const std::string& iconName, const QString& toolTip);
```

File: src/gui/SubWindow.cpp [hunk 5]
```diff
@@ -111,12 +95,17 @@ SubWindow::SubWindow(QWidget *parent, Qt::WindowFlags windowFlags) :
  */
 void SubWindow::paintEvent( QPaintEvent * )
 {
+	// Don't paint any of the other stuff if the sub window is maximized
+	// so that only its child content is painted.
+	if (isMaximized())
```
Comment (Rossmaxx) [inline src/gui/SubWindow.cpp:100]: Would it make sense to inline this block?

File: src/gui/SubWindow.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (JohannesLorenz) [reply src/gui/SubWindow.cpp:296]: That code change is OK.

File: include/SubWindow.h (hunk 4 above)
Comment (michaelgregorius) [reply include/SubWindow.h]: Fixed with commit e89694a2199.

File: src/gui/SubWindow.cpp (hunk 5 above)
Comment (Rossmaxx) [reply src/gui/SubWindow.cpp:100]: I mean put the return statement in the same line as the if, with braces.

Comment (michaelgregorius) [reply src/gui/SubWindow.cpp]: Ah, ok. Done with commit d5be2e5e6ad. For some reasons I had the code of `adjustTitleBar` in mind when reading the question. That method has a bit more code in the scope of the `if (isMaximized())` statement.

**Comment by michaelgregorius** [comment]
@JohannesLorenz, @Rossmaxx, I have merged this PR so that I can in turn merge `master` back into #7514 (done with commit f677b8c8cb0).

**Comment by JohannesLorenz** [comment]
Very good, that makes #7514 much more readable!

**Comment by michaelgregorius** [comment]
> Very good, that makes #7514 much more readable!

Yes, due to splitting it up into two PRs #7514 now is only about giving instrument windows the ability to maximize.

**Pull request merged**